### PR TITLE
Build tooling upgrade and simplification

### LINKED ISSUE #765: Remove unnecessary nodejs fileReader helper function
This will allow us to avoid the extra dep on fs, which may cause issues for some bundlers.

### PR DESCRIPTION
**Pull request by carsonfarmer**
## Description

This is a big administrative PR. It includes some major speedups to clean, bootstrap, and build times. It reduces extraneous deps throughout the sub-packages, merges deps at the top-level of the repo, and takes better advantage of npx caching. It is also a precursor to ESM builds.

This should NOT change any APIs or externally exported types. The only "API" change is that this also removes a helper API that we only just recently added (see below).

Fixes #765 

The most significant change for devs is that the build command is no longer automatically run upon install. This should speed up bootstrapping significantly, and makes releases easier (though slightly more manual). If you use the `npm run release` script, you should be fine and no additional changes to your build process are required. However, editing locally, you might need to use the tsc watch functionality, or just run `npm run build` explicitly if you are making significant changes. You should not have to run `npm run bootstrap` as frequently as before, and actually, the following patter should be pretty quick:

```
npm run clean
npm run bootstrap
npm run build
```

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

No tests were changed (save for some linting and organizational changes).

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -88,5 +75,56 @@
     "typedoc": "0.17.7",
     "typescript": "^3.7.2",
     "typescript-docs-verifier": "^1.1.2"
+  },
```
Comment (carsonfarmer) [inline package.json:78]: pulled all the dot files into here to clean things up a bit

Comment (carsonfarmer) [reply package.json:78]: note also that there are a lot of "new" and existing linter issues left in this repo, so we'll have to make sure we all update our IDE's linter settings to use the stuff in this repo

File: packages/buckets/src/api/index.ts [hunk 2]
```diff
@@ -53,9 +53,9 @@ import {
 import { Context, ContextInterface } from '@textile/context'
 import { GrpcConnection } from '@textile/grpc-connection'
 import { WebsocketTransport } from '@textile/grpc-transport'
-import CID from 'cids'
 import { EventIterator } from 'event-iterator'
 import log from 'loglevel'
+import { CID } from 'multiformats'
```
Comment (carsonfarmer) [inline packages/buckets/src/api/index.ts:58]: Switching to multiformity over cids due to some release regressions. This should be low impact, but if it is an issue, we can revert. It _will_ require a minor version bump. The change_should_ be backward compatible, but it is an external dev change, so we might want to consider just doing a major bump to be on the safe side.

File: packages/buckets/src/api/index.ts [hunk 3]
```diff
@@ -216,8 +218,13 @@ async function ensureRootString(
   }
 }
 
-export function* genChunks(value: Uint8Array, size: number) {
-  return yield* Array.from(Array(Math.ceil(value.byteLength / size)), (_, i) => value.slice(i * size, i * size + size))
+export function* genChunks(
```
Comment (carsonfarmer) [inline packages/buckets/src/api/index.ts:221]: Just linter stuff here

File: packages/buckets/src/api/index.ts [hunk 4]
```diff
@@ -405,11 +423,15 @@ export async function bucketsPushPath(
   input: any,
   opts?: PushOptions,
   ctx?: ContextInterface,
-) {
+): Promise<PushPathResult> {
```
Comment (carsonfarmer) [inline packages/buckets/src/api/index.ts:426]: We have been lax with these. Our docs building will be happy with use for this. I've added this as a linter check, so it should be easier to keep this up to date.

File: packages/buckets/src/api/index.ts [hunk 5]
```diff
@@ -505,7 +529,7 @@ export async function bucketsPushPathNode(
...
```
Comment (carsonfarmer) [inline packages/buckets/src/api/index.ts:532]: Re: Above, our linter will now try to enforce explicitly defined types on external APIs.

File: packages/buckets/src/api/index.ts [hunk 6]
```diff
@@ -656,13 +685,18 @@ export function bucketsPullPath(
...
+        message: string,
+        // _trailers: grpc.Metadata,
+      ) => {
         if (status !== grpc.Code.OK) {
           fail(new Error(message))
         }
         stop()
       },
     })
+    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```
Comment (carsonfarmer) [inline packages/buckets/src/api/index.ts:699]: This is related to the above new linter setting. We don't always need to be strict if we have a well-defined response like this.

File: packages/buckets/src/api/normalize.ts [hunk 7]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
```
Comment (carsonfarmer) [inline packages/buckets/src/api/normalize.ts:1]: Just linter stuff here

File: packages/buckets/src/buckets.spec.ts [hunk 8]
```diff
@@ -477,9 +490,13 @@ describe('Buckets...', function () {
 
       const { root } = await bobBuckets.listPath(rootKey, '')
       try {
-        // Defaults to highwatermark of CHUNK_SIZE
-        const stream = createReadStream(path.join(pth, 'file2.jpg'))
-        await bobBuckets.pushPath(rootKey, 'path/to/bobby.jpg', stream, { root })
+        // Use default highwatermark of CHUNK_SIZE!
+        const stream = fs.createReadStream(path.join(pth, 'file2.jpg'), {
```
Comment (carsonfarmer) [inline packages/buckets/src/buckets.spec.ts:494]: Revert to using default createReadStream from node. We don't export a helper function anymore, it wasn't worth the extra deps.

File: packages/buckets/src/buckets.ts [hunk 9]
```diff
@@ -144,7 +144,10 @@ export class Buckets extends GrpcAuthentication {
    * }
```
Comment (carsonfarmer) [inline packages/buckets/src/buckets.ts:144]: linter and explicit return types

File: packages/buckets/src/spec.util.ts [hunk 10]
```diff
@@ -1,11 +1,15 @@
...
+import {
+  APIServiceClient,
+  ServiceError,
+} from '@textile/hub-grpc/api/hubd/pb/hubd_pb_service'
 import axios from 'axios'
-import delay from 'delay'
-import { HMAC } from 'fast-sha256'
-import multibase from 'multibase'
+
+const delay = (time: number): Promise<void> =>
```
Comment (carsonfarmer) [inline packages/buckets/src/spec.util.ts:11]: no need for a whole separate package for something like this

File: packages/buckets/src/types.ts [hunk 11]
```diff
@@ -1,4 +1,4 @@
-import CID from 'cids'
+import type { CID } from 'multiformats'
```
Comment (carsonfarmer) [inline packages/buckets/src/types.ts:1]: we only need the types here. this is safer when building as we can tree-shake the whole dep out!

File: packages/crypto/package.json [hunk 12]
```diff
@@ -12,35 +12,21 @@
...
-		"build": "npm run build:protos && npm run build:compile",
-		"build:compile": "tsc -b tsconfig.json",
+		"build": "npm run build:protos && npm run build:cjs",
+		"build:cjs": "npx tsc -b ./tsconfig.json",
 		"build:protos": "npx pbjs src/keys.proto --ts src/proto.keys.ts",
-		"clean": "rimraf ./dist ./tsconfig.tsbuildinfo",
+		"clean": "npx rimraf ./dist ./tsconfig.tsbuildinfo",
 		"test:browser": "polendina --cleanup --timeout 60 src/**/*.spec.ts --webpack-config ../../webpack.test.js"
 	},
-	"repository": {
```
Comment (carsonfarmer) [inline packages/crypto/package.json:24]: uniform package.json setup, plus cleanup references. all package.json got this treatment

File: packages/hub-filecoin/package.json [hunk 13]
```diff
@@ -39,24 +29,5 @@
     "@textile/security": "^0.6.2",
     "event-iterator": "^2.0.0",
     "loglevel": "^1.6.8"
-  },
-  "devDependencies": {
```
Comment (carsonfarmer) [inline packages/hub-filecoin/package.json:43]: bunch copy-pasta deps that aren't needed here. also references are now auto-added to tsconfig

File: packages/hub-filecoin/package.json [hunk 14]
```diff
@@ -12,21 +12,11 @@
     "access": "public"
   },
   "scripts": {
-    "prepublishOnly": "npm run build",
-    "prepare": "npm run build",
-    "prebuild": "npm run clean",
-    "build": "tsc -b tsconfig.json",
-    "clean": "rimraf ./dist ./tsconfig.tsbuildinfo",
+    "build": "npx tsc -b ./tsconfig.json",
```
Comment (carsonfarmer) [inline packages/hub-filecoin/package.json:15]: we can remove a ton of extra deps by sticking with npx here

File: packages/hub-filecoin/tsconfig.json [hunk 15]
```diff
@@ -2,11 +2,11 @@
   "extends": "../../tsconfig.json",
   "compilerOptions": {
     "outDir": "./dist",
-    "rootDir": "./src",
-    "resolveJsonModule": true
+    "rootDir": "./src"
   },
-  "exclude": [
-    "node_modules",
-    "dist"
```
Comment (carsonfarmer) [inline packages/hub-filecoin/tsconfig.json:10]: these are now auto added. leads to faster build times, and easier maintenance

File: packages/hub-filecoin/tsconfig.json [hunk 16]
```diff
@@ -2,11 +2,11 @@
   "extends": "../../tsconfig.json",
   "compilerOptions": {
     "outDir": "./dist",
-    "rootDir": "./src",
```
Comment (carsonfarmer) [inline packages/hub-filecoin/tsconfig.json:5]: now pretty much just mirroring our top-level tsconfig

File: packages/hub-threads-client/package.json [hunk 17]
```diff
@@ -12,21 +12,11 @@
     "access": "public"
   },
   "scripts": {
-    "prepublishOnly": "npm run build",
```
Comment (carsonfarmer) [inline packages/hub-threads-client/package.json:15]: you'll notice these are all different now. we _don't_ automatically try to rebuild on install. build is now an explicit step, that is _required_ before a publish/release. If you use the top-level release script, it'll do it automatically, so do that. But this also means you'll have to "manually" call npm run build if you make significant changes during build time. the good news is you can use watch, and the build will be much faster.

File: packages/hub-threads-client/src/client.spec.ts [hunk 18]
```diff
@@ -1,12 +1,12 @@
-import { ThreadID } from '@textile/threads-id'
 import { grpc } from '@improbable-eng/grpc-web'
+import { Context } from '@textile/context'
```
Comment (carsonfarmer) [inline packages/hub-threads-client/src/client.spec.ts:2]: linter

File: packages/hub/src/index.ts [hunk 19]
```diff
@@ -2,11 +2,11 @@
...
  */
-export * from '@textile/users'
 export * from '@textile/buckets'
 export * from '@textile/crypto'
 export * from '@textile/grpc-authentication'
 export * from '@textile/hub-filecoin'
 export * from '@textile/hub-threads-client'
 export * from '@textile/security'
 export { ThreadID, Variant } from '@textile/threads-id'
+export * from '@textile/users'
```
Comment (carsonfarmer) [inline packages/hub/src/index.ts:12]: linter, is this "safe" @andrewxhill ?

File: tsconfig.json [hunk 20]
```diff
@@ -7,33 +7,31 @@
       "dom",
```
Comment (carsonfarmer) [inline tsconfig.json:7]: simplification and better cross-package defaults

**Review by carsonfarmer — COMMENTED** [review]
Some guiding comments for you @andrewxhill and @asutula

File: packages/hub/src/index.ts (hunk 19 above)
Comment (andrewxhill) [reply packages/hub/src/index.ts:12]: which part of it?

**Review by andrewxhill — APPROVED** [review]

File: CODE_OF_CONDUCT.md [hunk 21]
```diff
@@ -3,33 +3,33 @@
 ## Our Pledge
```
Comment (andrewxhill) [inline CODE_OF_CONDUCT.md:3]: do we want to move this into docs repo with a single reference for all other repos? docs.textile.io/terms/... is where others exist

Comment (carsonfarmer) [reply CODE_OF_CONDUCT.md:3]: yeh that's probably a good idea! i just copied the updates from go-threads actually, but we might as well just link to a single source of truth in all repos and be done with it.

File: packages/hub/src/index.ts (hunk 19 above)
Comment (carsonfarmer) [reply packages/hub/src/index.ts:12]: that the linter moved export * from users to the bottom

**Pull request merged**